UnderlineNav with updated design for states and counter

broccolinisoup opened the pull request:
Closes [#1240](https://github.com/github/primer/issues/1240)

UnderlineNav has updated design and this PR is for the updated states (default, selected, hover and focus state) as well as adding counter as a prop.

### Screenshots 

<img width="614" alt="Screen Shot 2022-09-01 at 3 13 03 pm" src="https://user-images.githubusercontent.com/1446503/187836663-84c28dba-0143-4b0c-9772-9ae53b0b83e4.png">
<img width="590" alt="Screen Shot 2022-09-01 at 3 12 32 pm" src="https://user-images.githubusercontent.com/1446503/187836666-8ed15a2b-4a38-48b6-a774-18ef6ca261b9.png">

### Screen Recording


https://user-images.githubusercontent.com/1446503/187836734-e7d1c516-5ab2-445c-83d4-bf20b9995f51.mp4


### Merge checklist

- [x] Added/updated tests
- [x] Added/updated documentation
- [x] Tested in Chrome
- [x] Tested in Firefox
- [x] Tested in Safari
- [x] Tested in Edge

Take a look at the [What we look for in reviews](https://github.com/primer/react/blob/main/contributor-docs/CONTRIBUTING.md#what-we-look-for-in-reviews) section of the contributing guidelines for more information on how we review PRs.

broccolinisoup commented on src/UnderlineNav2/UnderlineNavItem.tsx:
```diff
@@ -70,30 +77,51 @@ export const UnderlineNavLink = forwardRef(
       whiteSpace: 'nowrap'
     }
+
+    const wrapperStyles = {
```
I added another Box (div) that wraps icon, text and the counter and gave some padding to achieve smaller grey rectangular hover state.
<img width="181" alt="Screen Shot 2022-08-29 at 3 12 39 pm" src="https://user-images.githubusercontent.com/1446503/187127575-950233ea-ae0d-42d3-8596-1f361e086b1c.png">

broccolinisoup commented on src/UnderlineNav2/UnderlineNavItem.tsx:
```diff
@@ -70,30 +77,51 @@ export const UnderlineNavLink = forwardRef(
+      paddingX: 2
+    }
     const smallVariantLinkStyles = {
-      paddingY: 2,
```
reduced the paddingY as a unit comes from the wrapper

broccolinisoup commented on src/UnderlineNav2/UnderlineNavItem.tsx:
```diff
@@ -70,30 +77,51 @@ export const UnderlineNavLink = forwardRef(
       ...(variant === 'small' ? smallVariantLinkStyles : defaultVariantLinkStyles),
-      '&:hover, &:focus': {
-        borderColor: selectedLink === ref ? 'primer.border.active' : 'neutral.muted',
+      '&:hover > div[data-component="wrapper"] ': {
```
when anchor is hover, I added styles to the wrapper.

broccolinisoup commented on src/UnderlineNav2/UnderlineNavItem.tsx:
```diff
@@ -70,30 +77,51 @@ export const UnderlineNavLink = forwardRef(
       display: 'inline-flex',
       color: 'fg.default',
       textAlign: 'center',
-      borderBottom: '2px solid transparent',
```
I had to remove the border entirely because of the focus style below. (The focus class on link element)

     &:focus': {
        outlineColor: 'accent.fg',
        borderRadius: 2,
        outlineOffset: '-6px',
        transition: '0.2s ease'
      }

The outline border can only be rounded with borderRadius. However when I added the borderRadius to the focus style, as this focus class gets triggered when **the item is selected as well**, it ended up rounding the orange border bottoms too. So I removed the borders from the links and add a standalone element and styled them to look like border bottom so that they won't affect from the focus styles. Please let me know what you think or if you have a different idea.

broccolinisoup replied:
with the new commit this is not relevant anymore

broccolinisoup commented on src/UnderlineNav2/UnderlineNavItem.tsx:
```diff
@@ -129,18 +157,26 @@ export const UnderlineNavLink = forwardRef(
+            )}
+          </Box>
         </Box>
+        {selectedLink === ref && <Box as="span" sx={borderStyles}></Box>}
```
Standalone element to represent the border bottom.

broccolinisoup replied:
with the new commit this is not relevant anymore

broccolinisoup commented on src/UnderlineNav2/UnderlineNavItem.tsx:
```diff
@@ -70,30 +77,51 @@ export const UnderlineNavLink = forwardRef(
+
+    const borderStyles = {
+      // How to use primer primitives for space 3?
+      width: `calc(100% - 16px)`,
```
How to use primer primitives for space 3?

broccolinisoup commented on src/UnderlineNav2/UnderlineNavItem.tsx:
```diff
@@ -70,30 +77,51 @@ export const UnderlineNavLink = forwardRef(
+      '&:focus': {
+        outlineColor: 'accent.fg',
+        borderRadius: 2,
+        outlineOffset: '-6px',
```
I don't think we can set X and Y outline offsets ([ref](https://developer.mozilla.org/en-US/docs/Web/CSS/outline-offset)) , so this is the limitation came from using native outline.  (Not super close to the design -  see the x paddings are less in the design)

### Design
<img width="197" alt="Screen Shot 2022-08-29 at 3 33 57 pm" src="https://user-images.githubusercontent.com/1446503/187129844-9a9c9bf5-7884-4d80-beff-b866c0bc9b0c.png">

### Storybook
<img width="190" alt="Screen Shot 2022-08-29 at 3 40 39 pm" src="https://user-images.githubusercontent.com/1446503/187130582-b0343149-ccce-4b2b-b357-6d1693af0fa0.png">

broccolinisoup replied:
with the new commit this is not relevant anymore

pksjce commented on src/UnderlineNav2/index.ts:
```diff
@@ -1,10 +1,10 @@
 
 const UnderlineNav = Object.assign(Nav, {
-  Link: UnderlineNavLink
+  Item: UnderlineNavItem
```
This would be a breaking change wrt to the old UnderlineNav. We will need to mention it in the changelog.

pksjce reviewed (commented):
The focus rings work so well! 💯 

I can see two things out of place.
1. The underline from the nav seems to have gone away.
2. The selected underline seems a bit longer on both sides compared to the Item.

![image](https://user-images.githubusercontent.com/417268/187349266-56525337-eb58-4a13-8e35-e5aa94a6aba1.png)

Also I'm sceptical about the name change from UnderlineNav.Link to UnderlineNav.Item. It would break existing API. Though, I agree its a change that makes full sense. Not sure how to proceed with it honestly.

broccolinisoup commented:
Thanks for the review @pksjce ! 

> The underline from the nav seems to have gone away.

I used [the states reference](https://www.figma.com/file/N9JSJxBe6XAIeUjkvl4Bsy/UnderlineNav-Updates-Q1?node-id=8%3A12051) from Figma and there is no underline there so I though this line must have been removed with the updated design but for all the other references in Figma, the line is there! I added it back ✨  Sorry for the confusion! 

> The selected underline seems a bit longer on both sides compared to the Item.

Yes, you are right! I might have missed that. I also realised there was way too much margin between the li items. So all together, I re-styled them and this is for the reference to compare before and after. (I updated the PR description as well with the new screenshots)

Before

<img width="660" alt="Screen Shot 2022-08-29 at 3 04 21 pm" src="https://user-images.githubusercontent.com/1446503/187362597-03f95ccf-3b9b-4138-835c-1effb9074789.png">

After
<img width="581" alt="Screen Shot 2022-08-30 at 4 11 22 pm" src="https://user-images.githubusercontent.com/1446503/187362755-de433271-6cdb-4b5c-85e7-74b8d063911b.png">

And the design

<img width="566" alt="Screen Shot 2022-08-30 at 4 14 24 pm" src="https://user-images.githubusercontent.com/1446503/187363261-950bf220-bfdc-49b5-912a-2093e47d8bdd.png">

> Also I'm sceptical about the name change from UnderlineNav.Link to UnderlineNav.Item. It would break existing API. Though, I agree its a change that makes full sense. Not sure how to proceed with it honestly.

Yeah I agree. It makes sense to me too and I see other "new" components such as ActionList uses Item too so it is good to be consistent. I guess we can discuss with others as well and find the best strategy to move forward with the major release.

danielguillan commented:
Sorry for the confusion about the removal of the underline. 

I initially intentionally removed it so with the removal of `actions,` consumers could easily build a custom UnderlineNav + actions horizontal layout and deal with the bottom border on the parent container. I also did so to support the use case in the code clone tab where we don't use a border at all.

But now that we have a clearer idea of what needs to be supported (and not) and how we'll move forward in the mid-long term, we should bring the underline back.

danielguillan commented:
I also think the name change from `UnderlineNav.Link` to `UnderlineNav.Item` makes sense. Please feel free to point anyone interested in this discussion to the design API proposal: https://github.com/github/primer/pull/1239

broccolinisoup commented on src/UnderlineNav2/UnderlineNavItem.tsx:
```diff
@@ -70,30 +77,73 @@ export const UnderlineNavLink = forwardRef(
+        backgroundColor: 'neutral.muted',
+        transition: 'background .12s ease-out'
+      },
+      '&:focus': {
```
new focus implementation. As @langermank pointed out today that Safari doesn't respect borderRadius for the outlines so new focus implementation, inspired from [Primer CSS](https://github.com/primer/css/blob/main/src/support/mixins/misc.scss)

broccolinisoup commented on src/UnderlineNav2/UnderlineNavItem.tsx:
```diff
@@ -70,30 +77,73 @@ export const UnderlineNavLink = forwardRef(
+        boxShadow: `inset 0 0 0 2px #0969da`
+      },
+      // renders a visibly hidden "copy" of the label in bold, reserving box space for when label becomes bold on selected
+      '& span[data-content]::before': {
```
From [Primer CSS](https://github.com/primer/css/blob/main/src/navigation/underline-nav.scss#L59) 🙌🏼

broccolinisoup commented on src/UnderlineNav2/UnderlineNavItem.tsx:
```diff
@@ -70,30 +77,73 @@ export const UnderlineNavLink = forwardRef(
+        visibility: 'hidden'
+      },
+      // selected state styles
+      '&::after': {
```
Previous element for the border bottom was creating a jumpy affect when selected as it was conditionally rendered. So to avoid that jumpiness, used pseudo class (ref: [Primer CSS](https://github.com/primer/css/blob/main/src/navigation/underline-nav.scss#L90))

broccolinisoup commented on src/UnderlineNav2/UnderlineNavItem.tsx:
```diff
@@ -70,30 +77,73 @@ export const UnderlineNavLink = forwardRef(
+      '&:focus': {
+        outline: 0,
+        '& > div[data-component="wrapper"]': {
+          boxShadow: `inset 0 0 0 2px #0969da`
```
TODO: use primitives here

broccolinisoup commented on src/UnderlineNav2/UnderlineNav.tsx:
```diff
@@ -143,14 +156,14 @@ export const UnderlineNav = forwardRef(
           {actions.length > 0 && (
             <ActionMenu>
-              <ActionMenu.Button>More</ActionMenu.Button>
+              <ActionMenu.Button sx={{m: 0}}>More</ActionMenu.Button>
```
to fix extra margin around the button on safari

pksjce replied:
wow, can you add a comment here about this?

pksjce commented on src/UnderlineNav2/UnderlineNavItem.tsx:
```diff
@@ -36,16 +38,21 @@ export type UnderlineNavLinkProps = {
+  /**
+   * Counter
+   */
+  counter?: number
```
@danielguillan - Wondering if we are going to support loading state for counters.

broccolinisoup replied:
Just making a comment here for the types. I added the counter prop with only number type just to make the Counter component redundant. We can add another type for the loading state if we are going to support 😊

danielguillan replied:
Yes, we are going to support loading states. We can add support incrementally in another PR if that makes things easier.

pksjce commented on .changeset/thirty-mayflies-travel.md:
```diff
@@ -0,0 +1,5 @@
+---
+'@primer/react': major
```
I don't think this should be a major change yet. We'll kind of block ourselves doing that. Once we make all the changes, we can graduate from drafts to main folder with a major change. Could you do that before merging?
